### ISSUE TITLE
Google Style: Incorrect ParenPad warning

### ISSUE THREAD

**Issue by cushon** [checkstyle/checkstyle#4294]
http://checkstyle.sourceforge.net/config_whitespace.html#ParenPad

The Google Java Style Guide requires a space after `;` [1], so the following warning is incorrect.

[1] https://google.github.io/styleguide/javaguide.html#s4.6.2-horizontal-whitespace

>After ,:; or the closing parenthesis ')' of a cast

```java
class T {
  {
    for (; ; ) {
      //
    }
    try (InputStream is = null; ) { // violation
      //
    }
  }
}
```

```
$ java -jar checkstyle-7.6-all.jar -c google_checks.xml T.java
[WARN] T.java:6:32: ')' is preceded with whitespace. [ParenPad]
```

**Comment by romani** [comment]
pre-assigned to @soon

**Comment by soon** [comment]
I am on it

**Comment by soon** [comment]
The `ParenPad` check now contains only `option` parameter, which could be either `SPACE` or `NOSPACE`. I want to add another option like `spaceAfterSemicolon` (or a list of tokens, which should be followed with a space), which will force a user to put a space between a semicolon and a right parenthesis. 

Is it acceptable, or there is another solution?

**Comment by romani** [comment]
AST tree:
```
$ java -jar checkstyle-7.6.1-all.jar -t TestClass.java
CLASS_DEF -> CLASS_DEF [1:0]
|--MODIFIERS -> MODIFIERS [1:0]
|--LITERAL_CLASS -> class [1:0]
|--IDENT -> TestClass [1:6]
`--OBJBLOCK -> OBJBLOCK [1:16]
    |--LCURLY -> { [1:16]
    |--INSTANCE_INIT -> INSTANCE_INIT [2:2]
    |   `--SLIST -> { [2:2]
    |       |--LITERAL_FOR -> for [3:4]
    |       |   |--LPAREN -> ( [3:8]
    |       |   |--FOR_INIT -> FOR_INIT [3:9]
    |       |   |--SEMI -> ; [3:9]
    |       |   |--FOR_CONDITION -> FOR_CONDITION [3:11]
    |       |   |--SEMI -> ; [3:11]
    |       |   |--FOR_ITERATOR -> FOR_ITERATOR [3:13]
    |       |   |--RPAREN -> ) [3:13]
    |       |   `--SLIST -> { [3:15]
    |       |       `--RCURLY -> } [5:4]
    |       |--LITERAL_TRY -> try [6:4]
    |       |   |--RESOURCE_SPECIFICATION -> RESOURCE_SPECIFICATION [6:8]
    |       |   |   |--LPAREN -> ( [6:8]
    |       |   |   |--RESOURCES -> RESOURCES [6:9]
    |       |   |   |   `--RESOURCE -> RESOURCE [6:9]
    |       |   |   |       |--MODIFIERS -> MODIFIERS [6:9]
    |       |   |   |       |--TYPE -> TYPE [6:9]
    |       |   |   |       |   `--IDENT -> InputStream [6:9]
    |       |   |   |       |--IDENT -> is [6:21]
    |       |   |   |       `--ASSIGN -> = [6:24]
    |       |   |   |           `--EXPR -> EXPR [6:26]
    |       |   |   |               `--LITERAL_NULL -> null [6:26]
    |       |   |   |--SEMI -> ; [6:30]
    |       |   |   `--RPAREN -> ) [6:32]
    |       |   `--SLIST -> { [6:34]
    |       |       `--RCURLY -> } [8:4]
    |       `--RCURLY -> } [9:2]
    `--RCURLY -> } [10:0]
```

ParenPad is used in default mode in config - https://github.com/checkstyle/checkstyle/blob/master/src/main/resources/google_checks.xml#L177

http://checkstyle.sourceforge.net/config_whitespace.html#ParenPad
default tokens are LITERAL_FOR and RESOURCE_SPECIFICATION.

@soon , Why we violation only on  RESOURCE_SPECIFICATION ?

**Comment by soon** [comment]
@romani The ParenPad does not check right parenthesis in the for loop if the iterator part is empty (https://github.com/checkstyle/checkstyle/blob/888f258e2b14637cd7447dc4dbecc1c96d711109/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/ParenPadCheck.java#L171). The javadoc refers to the EmptyForIteratorPad check (https://github.com/checkstyle/checkstyle/blob/ce21086e087661553f3a774c38362327ee88996a/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyForIteratorPadCheck.java)

**Comment by romani** [comment]
ParenPad Check is not able to cover specific nuances language constructions with delimiters inside and optional parts.
Unfortunately in try-with-resources, it is only one part, last separation by ";"  is not required so user can put it or skip it..... that is why we have a problem here.

@cushon , what is a reason to keep extra `;` in `try` ?

**Comment by romani** [comment]
Looks like we have two ways to resolve nuance:
1) create special Check that will cover all nuances of try-with-resources as we already have for FOR structure.
2) some changes in ParenPad  ..... let me think

**Comment by cushon** [comment]
> what is a reason to keep extra `;` in try ?

I'm not aware of any reasons to prefer the extra `;`. When we see instances of this bug I've been recommending removing it as a work-around.

But it's valid Java code that shows up in the wild, and there are no plans to add a special case to the style guide for this, so it would be nice if checkstyle didn't warn about it.

**Comment by romani** [comment]
@rnveach  and @Vladlis (as you are mentioned in authors of this check)
we already got PR , but we did not decided how it should be resolved.

please review my comment https://github.com/checkstyle/checkstyle/issues/4294#issuecomment-299725656 with explanation of problem and possible scenarios to resolve.
The more I look to this the more I understand that try-with-resources is the same special braced element of java. In all other such java elements, trailing separator is not allowed at all. 
I more in favor of option 1).

Please share ides or vote.

**Comment by Vladlis** [comment]
@romani 
I like the first option.

**Comment by romani** [comment]
fix is merged.
